documentation/general/torn-sdk.md

```diff
@@ -0,0 +1,109 @@
+
+This also means these analyzers can vastly simplify their development model by alwasy targeting the latest version of Roslyn. There is no more need to multi-target beacuse the version of the compiler the analyzer will be used with is known at ship time for all core build scenarios.
+
+This means that our design time experience can differ from our command line experience when customers are in a torn state. Specifically it's possible, even likely in some cases, that the diagonstics produced by design time builds will differ from command line builds. That is a trade off that we are willing to make. Customers who wish to have a consistent experience between design time should not operate in a torn state.
+
+#### NuGet based analyzers
+
+Analyzers which ship via NuGet will continue to following the existing [support policies][matrix-of-paine]. This means that they will either need to target a sufficiently old version of Roslyn or implement multi-targeting support.
+
+## Work Breakdown
+
+- [ ]: Flow the Micosoft.Net.Compilers.Toolset.Framework package to the .NET SDK
+- [ ]: Create a new package Micosoft.NETSdk.Compiler.Roslyn in .NET SDK.
+  - [ ]: The contents of this package will include the contents of the Microsoft.Net.Compilers.Toolset.Framework package
+  - [ ]: The contents will include a README.md stating the package is **not** supported for direct user consumption.
+  - [ ]: The package will follow the versioning scheme of the .NET SDK.
+- [ ]: Change the Sdk.targets file to have copies of the following three `<UsingTasks>` from [Microsoft.Common.tasks][microsoft-common-tasks]. Having a copy in Sdk.targets means that resetting `$(RoslynTargetsPath)` during build will change the chosen compiler.
+- [ ]: When the .NET SDK detects a torn state
+  - [ ]: Use a `<PackageDownload>` to acquire the Microsoft.NETSdk.Compiler.Roslyn package
+  - [ ]: Change `$(RoslynTargetsPath)` to point into the package contents
+
+## Alternative
+
+### PackageReference
+
+Instead of `<PackageDownload>` the toolset package could be installed via `<PackageReference>`. This is how the existing Microsoft.Net.Compilers.Toolset package works today. This has a number of disadvantages:
+
+1. PackageReferences are not side effect free and have been shown to cause issues in complex builds. For example this package did not work by default in the Visual Studio or Bing builds. Deploying this at scale to all our customers will almost ceratinly cause issues.
+2. This approach does not solve all scenarios. The mechanics of restore mean that the compiler is not swapped until part way through the build. Build operations that happen earlier, such as legacy WPF, will still use the Visual Studio compiler and hence experience analyzer compatibility issues.
```

Comment:
> Build operations that happen earlier, such as legacy WPF, will still use the Visual Studio compiler and hence experience analyzer compatibility issues.

I'd love to learn more about this - WPF builds things before _Restore_?!

Reply:
@KirillOsenkov knows the details of why this occurs. I don't fully understand but I trust that he is correct on this.

Reply:
I believe it's not that it builds _before_ restore, it's that (older copies of) the WPF task/targets _don't include restore outputs_ and thus aren't affected by restore. https://github.com/dotnet/wpf/issues/9112 has more details if you were feeling too cheerful today.

Reply:
Yes, it's a mess. There can be three modes: 

 1. non-SDK (legacy) projects targeting desktop framework and using legacy WPF, 
 2. SDK-style projects targeting desktop framework and legacy WPF, and 
 3. SDK-style projects targeting modern .NET and modern open-source WPF (UseWPF=true). 

Only the third (modern) mode has the fixes to see PackageReferences and restore outputs. The first two import the targets from C:\Windows :(